### ISSUE TITLE
Move away from Gateway Configuration in Contour Config File

### ISSUE THREAD

**Issue by stevesloka** [projectcontour/contour#3617]
Currently, Contour requires a name/namespace of a Gateway to be [configured](https://github.com/projectcontour/contour/blob/main/examples/contour/01-contour-config.yaml#L15-L17) in the Contour configuration file. Today if this isn't configured and a user wants to use GatewayAPI, then they must set the value in the file, then restart Contour. 

We could move to use the `controller` field of the `GatewayClass` or a specific field in the `parametersRef`. This would allow the Gateway to be specified easily without needing a configuration file change / management.

**Comment by youngnick** [comment]
I think using the `controller` field in the GatewayClass is a better idea, and allows for more dynamic configuration. It's also the most Kube-y way to do it.

**Comment by stevesloka** [comment]
What would you use to specify the name/namespace? Something like: `projectcontour.io/name/namespace`? So for a gateway in the `projectcontour` namespace named `contour`, you'd have a GatewayClass like this:

```
kind: GatewayClass
apiVersion: networking.x-k8s.io/v1alpha1
metadata:
  name: contour-class
spec: 
  controller: projectcontour.io/contour/projectcontour
```

**Comment by youngnick** [comment]
I assumed that we'd just do like what we do with IngressClass, you specify some value, and the controller uses that, probably with a `projectcontour.io/` prefix. Doing things that way means that we need to handle what happens if more than one GatewayClass has the same `spec.controller` field, we've talked about that before.

But, this does highlight a similar problem to the IngressClass field/annotation: What happens if two Contour installations set the same IngressClass or GatewayClass `spec.controller`? I *think* that they'll probably fight right now?

For the simple internal-external multiple-Contours case, I think that having the namespace in there and not just some string (that is, `internal` and `external`) feels like overkill.

I'm not sure though, it feels like this bit could also use some upstream discussion maybe?

**Comment by stevesloka** [comment]
What I want to do here though is remove the need to configure Contour with something. So in the example of having an internal/external instances, you would configure the GatewayClass  to say, the Contour controller named `internal` in the namespace `ingress-internal` is who will service my routes for internal requests, then another GatewayClass would say, the Contour controller named `external` in the namespace `ingress-external` would service requests for external resources.

When deploying Contour, there's nothing special to configure, everything is driven from the GatewayClass and Contour doesn't need restarted for this to work.

**Comment by youngnick** [comment]
So, if Contour `internal` in `ingress-internal` is running before the GatewayClass already, then you don't need to configure a GatewayClass for it to be picked up? We *can* do it that way, sure.

But it seems weird to me. We already expect you to configure a IngressClass and restart Contour, why should this one be different? I feel like making GatewayClass easy to understand for new users is more important than cutting out a single Contour restart.

That is, I think it's easier to explain that "GatewayClass is IngressClass for Gateway APIs" is much, much easier to get than, "Create your GatewayClass with the following magic string and everything will just work".

**Comment by stevesloka** [comment]
This issue is not about the ingress class, this is defining the gateway/namespace that currently gets configured in the contour config file. 

I agree the magic string is not ideal which is why I also was thinking about the `parametersRef` section.

This issue is just a place for thoughts/discussion on how to make this easier. To me the less you need to configure/restart Contour the better. Things are much easier if the order of deployments doesn't matter, meaning, you could deploy Contour, deploy the CRDs, then create the GatewayClass and it should all work without having to restart things.

**Comment by youngnick** [comment]
> This issue is not about the ingress class, this is defining the gateway/namespace that currently gets configured in the contour config file.

Yes, I know. What I am saying is that I think that, if we're not going to specify the Gateway directly, then we should use the GatewayClass in the method it's intended, which is that you set your controller to look at all GatewayClasses, and find any that match, and use those. In our case, obviously, this is more complicated because a  single Contour installation can only handle one Gateway.

The more I type this, the more I remember the reasons why we suggested only doing a single one, because doing the full GatewayClass dance does not work well with our Contour == Gateway model.

> This issue is just a place for thoughts/discussion on how to make this easier.

I agree, and I think this is a good discussion!

> To me the less you need to configure/restart Contour the better. Things are much easier if the order of deployments doesn't matter, meaning, you could deploy Contour, deploy the CRDs, then create the GatewayClass and it should all work without having to restart things.

This, I don't agree with though. I think that applying a config change to a service is, for a the service owner, a *risk*. Having a possibly risky change applied automatically with no way to back out feels risky to me, and seems to be counter to the "Don't disrupt live traffic" thing that we are trying to build into Contour.

That is, having to restart for significant configuration changes is not a bug, it's a feature. When you enable the Gateway mode, you are significantly changing how Contour operates, and there should be a way to have the Contour owner need to opt-in to that behavior.

What happens if the Contour admin and the cluster operator are not the same person? Your whole Contour install can be affected if the cluster operator installs the Gateway CRDs, or configures a Gateway object without you knowing about it.

Requiring a restart puts a person in the loop, ensuring that someone either needs to take an action, or build a system to take the action for them. It provides backpressure to prevent cascading failure in the event that something goes really wrong.

All of this is, of course, dependent on what a "significant" config change is. But enabling new code paths seems significant enough to me.

**Comment by danehans** [comment]
Consider leaving the gateway ns/name config as-is and work on transitioning the contour configmap to a structured config using a CRD.

**Comment by danehans** [comment]
xref support for multiple controllers: https://github.com/projectcontour/contour/issues/3825

**Comment by skriss** [comment]
Closed by #3881
